canonical/loki-k8s-operator
# Add node exporter container to be able to plot log growth.

## Issue


This PR is necessary to fix #220.



## Solution


In this PR we added:

-  A [node-exporter](https://prometheus.io/docs/guides/node-exporter/#monitoring-linux-host-metrics-with-the-node-exporter) sidecar container that exposes a wide variety of hardware- and kernel-related metrics.



## Testing Instructions


1. Deploy COS-Lite: `juju deploy cos-lite --trust --channel=edge --overlay ./offers-overlay.yaml`
2. Deploy flog: `juju deploy flog-k8s flog --channel edge`
3. Configure flog to generate a big amount of logs: `juju config flog rate=25000`
4. Relate flog to Loki: `juju relate loki flog`
5. Pack the Loki charm: `charmcraft pack`
6. Refresh the Loki charm: `juju refresh loki --path ./*.charm --resource loki-image=ubuntu/loki:2-22.04 --resource node-exporter-image=prom/node-exporter:v1.7.0 --trust`
7. Check Prometheus has two endpoints for Loki, the first one is the endpoint Loki exposes, and the second one is the one node-exporter exposes:

   ![image](https://github.com/canonical/loki-k8s-operator/assets/939888/adada763-9afe-4d10-9c14-c9f2b19ed617)


## Bundle

```yaml
bundle: kubernetes
applications:
  alertmanager:
    charm: alertmanager-k8s
    channel: latest/edge
    revision: 108
    resources:
      alertmanager-image: 89
    scale: 1
    constraints: arch=amd64
    storage:
      data: kubernetes,1,1024M
    trust: true
  catalogue:
    charm: catalogue-k8s
    channel: latest/edge
    revision: 34
    resources:
      catalogue-image: 32
    scale: 1
    options:
      description: "Canonical Observability Stack Lite, or COS Lite, is a light-weight,
        highly-integrated, \nJuju-based observability suite running on Kubernetes.\n"
      tagline: Model-driven Observability Stack deployed with a single command.
      title: Canonical Observability Stack
    constraints: arch=amd64
    trust: true
  flog:
    charm: flog-k8s
    channel: latest/edge
    revision: 7
    resources:
      workload-image: 2
    scale: 1
    options:
      rate: 25000
    constraints: arch=amd64
  grafana:
    charm: grafana-k8s
    channel: latest/edge
    revision: 111
    resources:
      grafana-image: 68
      litestream-image: 43
    scale: 1
    constraints: arch=amd64
    storage:
      database: kubernetes,1,1024M
    trust: true
  loki:
    charm: local:loki-k8s-10
    scale: 1
    constraints: arch=amd64
    storage:
      active-index-directory: kubernetes,1,1024M
      loki-chunks: kubernetes,1,1024M
    trust: true
  prometheus:
    charm: prometheus-k8s
    channel: latest/edge
    revision: 175
    resources:
      prometheus-image: 142
    scale: 1
    constraints: arch=amd64
    storage:
      database: kubernetes,1,1024M
    trust: true
  traefik:
    charm: traefik-k8s
    channel: latest/edge
    revision: 178
    resources:
      traefik-image: 159
    scale: 1
    constraints: arch=amd64
    storage:
      configurations: kubernetes,1,1024M
    trust: true
relations:
- - traefik:ingress-per-unit
  - prometheus:ingress
- - traefik:traefik-route
  - grafana:ingress
- - traefik:ingress
  - alertmanager:ingress
- - prometheus:alertmanager
  - alertmanager:alerting
- - grafana:grafana-source
  - prometheus:grafana-source
- - grafana:grafana-source
  - alertmanager:grafana-source
- - prometheus:metrics-endpoint
  - traefik:metrics-endpoint
- - prometheus:metrics-endpoint
  - alertmanager:self-metrics-endpoint
- - prometheus:metrics-endpoint
  - grafana:metrics-endpoint
- - grafana:grafana-dashboard
  - prometheus:grafana-dashboard
- - grafana:grafana-dashboard
  - alertmanager:grafana-dashboard
- - catalogue:ingress
  - traefik:ingress
- - catalogue:catalogue
  - grafana:catalogue
- - catalogue:catalogue
  - prometheus:catalogue
- - catalogue:catalogue
  - alertmanager:catalogue
- - loki:logging
  - flog:log-proxy
- - loki:metrics-endpoint
  - prometheus:metrics-endpoint
- - loki:grafana-source
  - grafana:grafana-source
- - grafana:grafana-dashboard
  - loki:grafana-dashboard
--- # overlay.yaml
applications:
  alertmanager:
    offers:
      alertmanager-karma-dashboard:
        endpoints:
        - karma-dashboard
        acl:
          admin: admin
  grafana:
    offers:
      grafana-dashboards:
        endpoints:
        - grafana-dashboard
        acl:
          admin: admin
  prometheus:
    offers:
      prometheus-receive-remote-write:
        endpoints:
        - receive-remote-write
        acl:
          admin: admin
      prometheus-scrape:
        endpoints:
        - metrics-endpoint
        acl:
          admin: admin
```
